SOSS/dms/soss_centroids.py

```diff
@@ -0,0 +1,377 @@
+    image_masked = np.where(mask | ~refpix_mask, np.nan, image)
+
+    # Compute and subtract the background level of each column.
+    col_bkg = np.nanpercentile(image_masked, 10, axis=0)
+    image_masked_bkg = image_masked - col_bkg
+
+    # Find centroid - first pass, use all pixels in the column.
+    # Normalize each column
+    with np.errstate(invalid='ignore'):
+        image_norm = image_masked_bkg / np.nanmax(image_masked_bkg, axis=0)
+
+    # Create 2D Array of pixel positions.
+    xpix = np.arange(dimx)
+    ypix = np.arange(dimy)
+    _, ygrid = np.meshgrid(xpix, ypix)
+
+    # CoM analysis to find initial positions using all rows.
+    with np.errstate(invalid='ignore'):
+        ytrace = np.nansum(image_norm*ygrid, axis=0)/np.nansum(image_norm, axis=0)
+
+    # Second pass - use a windowed CoM at the previous position.
+    halfwidth = 30 * yos
+    for icol in range(dimx):
+
+        ycom = center_of_mass(image_norm[:, icol], ytrace[icol], halfwidth)
+
+        # If NaN was returned we are done.
+        if not np.isfinite(ycom):
+            ytrace[icol] = np.nan
+            continue
+
+        # If the pixel at the centroid is below the local mean we are likely mid-way between orders and
+        # we should shift the window downward to get a reliable centroid for order 1.
+        irow = np.int(np.around(ycom))
+        miny = np.int(np.fmax(np.around(ycom) - halfwidth, 0))
+        maxy = np.int(np.fmin(np.around(ycom) + halfwidth + 1, dimy))
+        if image_norm[irow, icol] < np.nanmean(image_norm[miny:maxy, icol]):
+            ycom = center_of_mass(image_norm[:, icol], ycom - halfwidth, halfwidth)
+
+        # If NaN was returned or the position is too close to the array edge, use NaN.
+        if not np.isfinite(ycom) or (ycom <= 5 * yos) or (ycom >= (ynative - 6) * yos):
+            ytrace[icol] = np.nan
+            continue
+
+        # Update the position if the above checks were succesfull.
+        ytrace[icol] = ycom
+
+    # Third pass - fine tuning using a smaller window.
+    halfwidth = 16 * yos
+    for icol in range(dimx):
+
+        ytrace[icol] = center_of_mass(image_norm[:, icol], ytrace[icol], halfwidth)
+
+    # Fit the y-positions with a polynomial and use the result as the true y-positions.
+    xtrace = np.arange(dimx)
+    mask = np.isfinite(ytrace)
+
+    # For padded arrays ignore padding for consistency with real data
+    if padding != 0:
+        mask = mask & (xtrace >= xos*padding) & (xtrace < (dimx - xos*padding))
+
+    # If no polynomial order was given return the raw measurements.
+    if poly_order is None:
+        param = []
+    else:
+        param = robust_polyfit(xtrace[mask], ytrace[mask], poly_order)
+        ytrace = np.polyval(param, xtrace)
+
+    # If verbose visualize the result.
+    if verbose is True:
+        _plot_centroid(image_masked, xtrace, ytrace)
+
+    return xtrace, ytrace, param
+
+
+def test_uncontam_centroids():
+    """"""
```

Comment:
Even if temporary have a docstring

If test function to remove late - put a `# TODO` comment before so we know it needs removal

Reply:
Removed.